trackio/ui/fns.py

```diff
@@ -49,10 +53,102 @@ def get_projects(request: gr.Request):
+        metrics_url += f"&write_token={write_token}"
+        runs_url += f"&write_token={write_token}"
+
     return gr.Navbar(
         value=[
-            ("Metrics", f"?selected_project={project_dd}"),
-            ("Runs", f"runs?selected_project={project_dd}"),
+            ("Metrics", metrics_url),
+            ("Runs", runs_url),
         ]
     )
+
+
+def check_hf_token_has_write_access(hf_token: str | None) -> None:
+    """
+    Checks to see if the provided hf_token is valid and has write access to the Space
+    that Trackio is running in. If the hf_token is valid or if Trackio is not running
+    on a Space, this function does nothing. Otherwise, it raises a PermissionError.
+    """
+    if os.getenv("SYSTEM") == "spaces":  # if we are running in Spaces
+        # check auth token passed in
+        if hf_token is None:
+            raise PermissionError(
+                "Expected a HF_TOKEN to be provided when logging to a Space"
+            )
+        who = HfApi.whoami(hf_token)
+        owner_name = os.getenv("SPACE_AUTHOR_NAME")
+        repo_name = os.getenv("SPACE_REPO_NAME")
+        # make sure the token user is either the author of the space,
+        # or is a member of an org that is the author.
+        orgs = [o["name"] for o in who["orgs"]]
+        if owner_name != who["name"] and owner_name not in orgs:
+            raise PermissionError(
+                "Expected the provided hf_token to be the user owner of the space, or be a member of the org owner of the space"
+            )
+        # reject fine-grained tokens without specific repo access
+        access_token = who["auth"]["accessToken"]
+        if access_token["role"] == "fineGrained":
+            matched = False
+            for item in access_token["fineGrained"]["scoped"]:
+                if (
+                    item["entity"]["type"] == "space"
+                    and item["entity"]["name"] == f"{owner_name}/{repo_name}"
+                    and "repo.write" in item["permissions"]
+                ):
+                    matched = True
+                    break
+                if (
+                    (
+                        item["entity"]["type"] == "user"
+                        or item["entity"]["type"] == "org"
+                    )
+                    and item["entity"]["name"] == owner_name
+                    and "repo.write" in item["permissions"]
+                ):
+                    matched = True
+                    break
+            if not matched:
+                raise PermissionError(
+                    "Expected the provided hf_token with fine grained permissions to provide write access to the space"
+                )
+        # reject read-only tokens
+        elif access_token["role"] != "write":
+            raise PermissionError(
+                "Expected the provided hf_token to provide write permissions"
+            )
+
+
+def check_oauth_token_has_write_access(oauth_token: str | None) -> None:
```

Comment:
The result from running whoami on an oauth token has a different format than when whoami is run on a regular hf token, hence the separate method.